File: src/src/main/resources/mmlspark/PageSplitter.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class PageSplitter(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        boundaryRegex (str): how to split into words (default: \s)
        inputCol (str): The name of the input column
        maximumPageLength (int): the maximum number of characters to be in a page (default: 5000)
        minimumPageLength (int): the the minimum number of characters to have on a page in order to preserve work boundaries (default: 4500)
        outputCol (str): The name of the output column (default: [self.uid]_output)
    """

    @keyword_only
    def __init__(self, boundaryRegex="\s", inputCol=None, maximumPageLength=5000, minimumPageLength=4500, outputCol=None):
        super(PageSplitter, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.PageSplitter")
        self.boundaryRegex = Param(self, "boundaryRegex", "boundaryRegex: how to split into words (default: \s)")
        self._setDefault(boundaryRegex="\s")
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.maximumPageLength = Param(self, "maximumPageLength", "maximumPageLength: the maximum number of characters to be in a page (default: 5000)")
        self._setDefault(maximumPageLength=5000)
        self.minimumPageLength = Param(self, "minimumPageLength", "minimumPageLength: the the minimum number of characters to have on a page in order to preserve work boundaries (default: 4500)")
        self._setDefault(minimumPageLength=4500)
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column (default: [self.uid]_output)")
        self._setDefault(outputCol=self.uid + "_output")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, boundaryRegex="\s", inputCol=None, maximumPageLength=5000, minimumPageLength=4500, outputCol=None):
        """
        Set the (keyword only) parameters

        Args:

            boundaryRegex (str): how to split into words (default: \s)
            inputCol (str): The name of the input column
            maximumPageLength (int): the maximum number of characters to be in a page (default: 5000)
            minimumPageLength (int): the the minimum number of characters to have on a page in order to preserve work boundaries (default: 4500)
            outputCol (str): The name of the output column (default: [self.uid]_output)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setBoundaryRegex(self, value):
        """

        Args:

            boundaryRegex (str): how to split into words (default: \s)

        """
        self._set(boundaryRegex=value)
        return self


    def getBoundaryRegex(self):
        """

        Returns:

            str: how to split into words (default: \s)
        """
        return self.getOrDefault(self.boundaryRegex)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setMaximumPageLength(self, value):
        """

        Args:

            maximumPageLength (int): the maximum number of characters to be in a page (default: 5000)

        """
        self._set(maximumPageLength=value)
        return self


    def getMaximumPageLength(self):
        """

        Returns:

            int: the maximum number of characters to be in a page (default: 5000)
        """
        return self.getOrDefault(self.maximumPageLength)


    def setMinimumPageLength(self, value):
        """

        Args:

            minimumPageLength (int): the the minimum number of characters to have on a page in order to preserve work boundaries (default: 4500)

        """
        self._set(minimumPageLength=value)
        return self


    def getMinimumPageLength(self):
        """

        Returns:

            int: the the minimum number of characters to have on a page in order to preserve work boundaries (default: 4500)
        """
        return self.getOrDefault(self.minimumPageLength)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column (default: [self.uid]_output)

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column (default: [self.uid]_output)
        """
        return self.getOrDefault(self.outputCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.PageSplitter"

    @staticmethod
    def _from_java(java_stage):
        module_name=PageSplitter.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".PageSplitter"
        return from_java(java_stage, module_name)

File: core/src/main/scala/com/microsoft/azure/synapse/ml/featurize/text/PageSplitter.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.featurize.text

import com.microsoft.azure.synapse.ml.codegen.Wrappable
import com.microsoft.azure.synapse.ml.core.contracts.{HasInputCol, HasOutputCol}
import com.microsoft.azure.synapse.ml.logging.{FeatureNames, SynapseMLLogging}
import org.apache.spark.injections.UDFUtils
import org.apache.spark.ml._
import org.apache.spark.ml.param._
import org.apache.spark.ml.util._
import org.apache.spark.sql.functions.col
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame, Dataset}

object PageSplitter extends DefaultParamsReadable[PageSplitter]

/** Splits text into chunks of at most n characters
  *
  * @param uid The id of the module
  */
class PageSplitter(override val uid: String)
  extends Transformer with HasInputCol with HasOutputCol
    with Wrappable with DefaultParamsWritable with SynapseMLLogging {
  logClass(FeatureNames.Featurize)

  def this() = this(Identifiable.randomUID("PageSplitter"))

  setDefault(outputCol, uid + "_output")

  val maximumPageLength =
    new IntParam(this, "maximumPageLength",
      "the maximum number of characters to be in a page")

  def setMaximumPageLength(v: Int): this.type = set(maximumPageLength, v)

  def getMaximumPageLength: Int = $(maximumPageLength)

  val minimumPageLength =
    new IntParam(this, "minimumPageLength",
      "the the minimum number of characters " +
        "to have on a page in order to preserve work boundaries")

  def setMinimumPageLength(v: Int): this.type = set(minimumPageLength, v)

  def getMinimumPageLength: Int = $(minimumPageLength)

  val boundaryRegex = new Param[String](this, "boundaryRegex", "how to split into words")

  def setBoundaryRegex(v: String): this.type = set(boundaryRegex, v)

  def getBoundaryRegex: String = $(boundaryRegex)

  setDefault(maximumPageLength -> 5000, minimumPageLength -> 4500, boundaryRegex -> "\\s")

  def split(textOpt: String): Seq[String] = {
    Option(textOpt).map { text =>
      if (text.length < getMaximumPageLength) {
        Seq(text)
      } else {
        val lengths = text
          .split(getBoundaryRegex)
          .map(_.length)
          .flatMap(l => List(l, 1))
          .dropRight(1)

        val indicies = lengths.scanLeft((0, 0, Nil: List[Int])) { case ((total, count, _), l) =>
          if (count + l < getMaximumPageLength) {
            (total + l, count + l, Nil)
          } else if (count > getMinimumPageLength) {
            (total + l, l, List(total))
          } else {
            val firstPageChars = getMaximumPageLength - count
            val firstPage = firstPageChars + total
            val remainingChars = l - firstPageChars

            val numPages = remainingChars / getMaximumPageLength
            val remainder = remainingChars - getMaximumPageLength * numPages
            val pages = List(firstPage) ::: (1 to numPages).map(i =>
              total + firstPageChars + getMaximumPageLength * i).toList
            (total + l, remainder, pages)
          }
        }.flatMap(_._3)

        val words = (List(0) ::: indicies.toList ::: List(text.length))
          .sliding(2)
          .map { case List(start, end) => text.substring(start, end) }
          .toSeq
        words
      }
    }.orNull
  }

  override def transform(dataset: Dataset[_]): DataFrame = {
    logTransform[DataFrame](
      dataset.toDF().withColumn(getOutputCol, UDFUtils.oldUdf(split _, ArrayType(StringType))(col(getInputCol))),
      dataset.columns.length
    )
  }

  override def copy(extra: ParamMap): PageSplitter =
    defaultCopy(extra)

  def transformSchema(schema: StructType): StructType = {
    assert(schema(getInputCol).dataType == StringType)
    schema.add(getOutputCol, ArrayType(StringType))
  }

}
